File: ios/Cindy/Health/Readiness.swift
import Foundation

/// How ready the body is for the next hard session, on a 0–100 scale.
///
/// Apple publishes no readiness figure — iOS 26 has no such HealthKit type —
/// so Cindy computes one from the values it can actually read and from its own
/// workout history. Every input is optional and the score is the weighted mean
/// of the components that exist, so the estimate degrades instead of
/// disappearing when there is no Apple Watch (or no Health access at all).
struct Readiness: Equatable {
    enum Band: String, Equatable {
        case rest, easy, ready, primed
    }

    /// One input, already translated to the common 0–100 scale.
    struct Component: Equatable, Identifiable {
        enum Kind: String, Equatable {
            case recovery, sleep, heartRateVariability, restingHeartRate, trainingLoad
        }

        var kind: Kind
        var score: Double
        var weight: Double
        /// Localized one-liner naming the measured value, e.g. "6.2 h of sleep last night".
        var detail: String

        var id: Kind { kind }
        /// How far this component pulls the total away from neutral.
        var pull: Double { (score - ReadinessScoring.neutral) * weight }
    }

    var score: Int
    var components: [Component]
    /// False when the score rests on the workout history alone.
    var usesHealthData: Bool

    var band: Band {
        switch score {
        case ..<40: return .rest
        case ..<60: return .easy
        case ..<80: return .ready
        default: return .primed
        }
    }

    /// Weighted score of everything except the plain hours-since-the-last-session
    /// rule: what the body says beyond what the calendar already knows.
    /// `nil` when recovery is the only thing known.
    ///
    /// The reminder uses this rather than the total, which already contains the
    /// waiting time it is about to add.
    var signalScore: Int? {
        let signals = components.filter { $0.kind != .recovery }
        let totalWeight = signals.reduce(0) { $0 + $1.weight }
        guard !signals.isEmpty, totalWeight > 0 else { return nil }
        return Int((signals.reduce(0) { $0 + $1.score * $1.weight } / totalWeight).rounded())
    }

    /// The components that move the score most, strongest first.
    func reasons(limit: Int = 3) -> [Component] {
        components.sorted { abs($0.pull) > abs($1.pull) }.prefix(limit).map { $0 }
    }

    /// `nil` when nothing at all is known — no workout in the history and no Health data.
    init?(components: [Component], usesHealthData: Bool) {
        let totalWeight = components.reduce(0) { $0 + $1.weight }
        guard !components.isEmpty, totalWeight > 0 else { return nil }
        let weighted = components.reduce(0) { $0 + $1.score * $1.weight }
        self.score = Int((weighted / totalWeight).rounded())
        self.components = components
        self.usesHealthData = usesHealthData
    }
}

/// The curves that turn each raw measurement into a 0–100 component score.
///
/// Deliberately piecewise linear and readable: these are judgement calls from
/// the training literature, not a validated model, and they should be easy to
/// argue with and to adjust.
enum ReadinessScoring {
    /// Score of an average day. Components sit around this when nothing stands out.
    static let neutral: Double = 78

    static let weights: [Readiness.Component.Kind: Double] = [
        .recovery: 0.35,
        .sleep: 0.20,
        .heartRateVariability: 0.25,
        .restingHeartRate: 0.10,
        .trainingLoad: 0.10,
    ]

    /// Rest taken since the last session, as a fraction of the rest it needs.
    static func recovery(fraction: Double) -> Double {
        interpolate(fraction, [(0, 10), (0.5, 45), (1, 85), (2, 100)])
    }

    /// Hours a session needs before the next hard one: two days for a full
    /// AMRAP, proportionally less for one that was cut short.
    static func recoveryHours(for record: WorkoutRecord) -> Double {
        let planned = max(record.plannedDuration, 1)
        let effort = min(1, record.durationSeconds / planned)
        return 24 + 24 * effort
    }

    static func sleep(hours: Double) -> Double {
        interpolate(hours, [(0, 0), (4, 25), (6, 60), (7.5, 88), (8.5, 100), (11, 85)])
    }

    /// Heart rate variability against the personal baseline: above is good.
    static func heartRateVariability(zScore: Double) -> Double {
        interpolate(zScore, [(-2.5, 5), (-1, 45), (0, 80), (1, 93), (2.5, 100)])
    }

    /// Resting heart rate against the personal baseline: above is bad.
    static func restingHeartRate(zScore: Double) -> Double {
        interpolate(-zScore, [(-2.5, 5), (-1, 45), (0, 80), (1, 93), (2.5, 100)])
    }

    /// Volume of the last 7 days divided by the weekly average of the last 28.
    /// Around 1 is sustainable; a sharp spike is the classic overreaching sign.
    static func trainingLoad(ratio: Double) -> Double {
        interpolate(ratio, [(0, 70), (0.5, 82), (0.8, 92), (1.0, 88), (1.3, 75), (1.6, 45), (2.5, 20)])
    }

    /// Linear interpolation between the given points, flat outside their range.
    static func interpolate(_ x: Double, _ points: [(Double, Double)]) -> Double {
        guard let first = points.first, let last = points.last else { return neutral }
        if x <= first.0 { return first.1 }
        if x >= last.0 { return last.1 }
        for (lower, upper) in zip(points, points.dropFirst()) where x <= upper.0 {
            let span = upper.0 - lower.0
            guard span > 0 else { return upper.1 }
            return lower.1 + (upper.1 - lower.1) * (x - lower.0) / span
        }
        return last.1
    }
}

/// A measurement together with the personal baseline it is judged against.
struct MetricBaseline: Equatable {
    var value: Double
    var mean: Double
    var standardDeviation: Double

    /// Deviation in standard deviations, clamped to ±3.
    /// `nil` when the baseline barely varies, where a z-score would explode.
    var zScore: Double? {
        guard standardDeviation > mean * 0.01, standardDeviation > 0 else { return nil }
        return min(max((value - mean) / standardDeviation, -3), 3)
    }

    /// Deviation from the baseline as a fraction, for the human-readable line.
    var relativeDeviation: Double {
        guard mean != 0 else { return 0 }
        return (value - mean) / mean
    }

    /// Baseline over earlier daily values, judged against the most recent one.
    /// `nil` below `minimumDays` of history, where "your normal" is guesswork.
    init?(current: Double, earlierDailyValues: [Double], minimumDays: Int = 10) {
        guard earlierDailyValues.count >= minimumDays else { return nil }
        let mean = earlierDailyValues.reduce(0, +) / Double(earlierDailyValues.count)
        let variance = earlierDailyValues.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / Double(earlierDailyValues.count)
        self.value = current
        self.mean = mean
        self.standardDeviation = variance.squareRoot()
    }

    init(value: Double, mean: Double, standardDeviation: Double) {
        self.value = value
        self.mean = mean
        self.standardDeviation = standardDeviation
    }
}

File: android/core/src/main/kotlin/me/raddatz/cindy/core/health/Readiness.kt
package me.raddatz.cindy.core.health

import me.raddatz.cindy.core.persistence.WorkoutRecord
import me.raddatz.cindy.core.roundedHalfAwayFromZero
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * How ready the body is for the next hard session, on a 0–100 scale.
 *
 * Neither Apple Health nor Health Connect publishes a readiness figure, so Cindy computes one
 * from the values it can actually read and from its own workout history. Every input is optional
 * and the score is the weighted mean of the components that exist, so the estimate degrades
 * instead of disappearing when there is no watch (or no health access at all).
 *
 * Create with [Readiness.from].
 */
@ConsistentCopyVisibility
data class Readiness private constructor(
    val score: Int,
    val components: List<Component>,
    /** False when the score rests on the workout history alone. */
    val usesHealthData: Boolean,
) {
    enum class Band { REST, EASY, READY, PRIMED }

    /** One input, already translated to the common 0–100 scale. */
    data class Component(
        val kind: Kind,
        val score: Double,
        val weight: Double,
        /** What was measured, for the one-liner the UI shows (iOS: a localized `detail` string). */
        val detail: ReadinessDetail? = null,
    ) {
        enum class Kind { RECOVERY, SLEEP, HEART_RATE_VARIABILITY, RESTING_HEART_RATE, TRAINING_LOAD }

        val id: Kind get() = kind

        /** How far this component pulls the total away from neutral. */
        val pull: Double get() = (score - ReadinessScoring.NEUTRAL) * weight
    }

    val band: Band
        get() = when {
            score < 40 -> Band.REST
            score < 60 -> Band.EASY
            score < 80 -> Band.READY
            else -> Band.PRIMED
        }

    /**
     * Weighted score of everything except the plain hours-since-the-last-session rule: what the
     * body says beyond what the calendar already knows. `null` when recovery is the only thing
     * known.
     *
     * The reminder uses this rather than the total, which already contains the waiting time it is
     * about to add.
     */
    val signalScore: Int?
        get() {
            val signals = components.filter { it.kind != Component.Kind.RECOVERY }
            val totalWeight = signals.fold(0.0) { sum, c -> sum + c.weight }
            if (signals.isEmpty() || !(totalWeight > 0)) return null
            return (signals.fold(0.0) { sum, c -> sum + c.score * c.weight } / totalWeight)
                .roundedHalfAwayFromZero().toInt()
        }

    /** The components that move the score most, strongest first. */
    fun reasons(limit: Int = 3): List<Component> =
        components.sortedByDescending { abs(it.pull) }.take(limit)

    companion object {
        /** `null` when nothing at all is known — no workout in the history and no health data. */
        fun from(components: List<Component>, usesHealthData: Boolean): Readiness? {
            val totalWeight = components.fold(0.0) { sum, c -> sum + c.weight }
            if (components.isEmpty() || !(totalWeight > 0)) return null
            val weighted = components.fold(0.0) { sum, c -> sum + c.score * c.weight }
            return Readiness(
                score = (weighted / totalWeight).roundedHalfAwayFromZero().toInt(),
                components = components,
                usesHealthData = usesHealthData,
            )
        }
    }
}

/** Typed replacement for the localized iOS `Readiness.Component.detail`. */
sealed interface ReadinessDetail {
    /** "\(hours) h since the last workout, \(needed) h suggested" (both rounded to whole hours). */
    data class SinceLastWorkout(val hoursSince: Int, val hoursSuggested: Int) : ReadinessDetail

    /** "This week's volume is \(ratio.formattedDecimal(1))× your average" */
    data class TrainingLoad(val ratio: Double) : ReadinessDetail

    /** "\(hours.formattedDecimal(1)) h of sleep last night" */
    data class Sleep(val hours: Double) : ReadinessDetail

    /**
     * "HRV \(value) ms, your average \(mean) ms" (rounded). Whatever HRV metric the platform
     * delivers — SDNN on iOS, RMSSD on Health Connect — both in milliseconds.
     */
    data class HeartRateVariability(val milliseconds: Int, val averageMilliseconds: Int) : ReadinessDetail

    /** "Resting pulse \(value) bpm, your average \(mean) bpm" (rounded). */
    data class RestingHeartRate(val beatsPerMinute: Int, val averageBeatsPerMinute: Int) : ReadinessDetail
}

/**
 * The curves that turn each raw measurement into a 0–100 component score.
 *
 * Deliberately piecewise linear and readable: these are judgement calls from the training
 * literature, not a validated model, and they should be easy to argue with and to adjust.
 */
object ReadinessScoring {
    /** Score of an average day. Components sit around this when nothing stands out. */
    const val NEUTRAL: Double = 78.0

    val weights: Map<Readiness.Component.Kind, Double> = mapOf(
        Readiness.Component.Kind.RECOVERY to 0.35,
        Readiness.Component.Kind.SLEEP to 0.20,
        Readiness.Component.Kind.HEART_RATE_VARIABILITY to 0.25,
        Readiness.Component.Kind.RESTING_HEART_RATE to 0.10,
        Readiness.Component.Kind.TRAINING_LOAD to 0.10,
    )

    /** Rest taken since the last session, as a fraction of the rest it needs. */
    fun recovery(fraction: Double): Double =
        interpolate(fraction, listOf(0.0 to 10.0, 0.5 to 45.0, 1.0 to 85.0, 2.0 to 100.0))

    /**
     * Hours a session needs before the next hard one: two days for a full AMRAP, proportionally
     * less for one that was cut short.
     */
    fun recoveryHours(record: WorkoutRecord): Double {
        val planned = maxOf(record.plannedDuration, 1.0)
        val effort = minOf(1.0, record.durationSeconds / planned)
        return 24 + 24 * effort
    }

    fun sleep(hours: Double): Double =
        interpolate(hours, listOf(0.0 to 0.0, 4.0 to 25.0, 6.0 to 60.0, 7.5 to 88.0, 8.5 to 100.0, 11.0 to 85.0))

    /**
     * Heart rate variability against the personal baseline: above is good. Only the z-score is
     * used, so it works for SDNN (iOS) and RMSSD (Health Connect) alike, as long as the baseline
     * is built from the same metric.
     */
    fun heartRateVariability(zScore: Double): Double =
        interpolate(zScore, listOf(-2.5 to 5.0, -1.0 to 45.0, 0.0 to 80.0, 1.0 to 93.0, 2.5 to 100.0))

    /** Resting heart rate against the personal baseline: above is bad. */
    fun restingHeartRate(zScore: Double): Double =
        interpolate(-zScore, listOf(-2.5 to 5.0, -1.0 to 45.0, 0.0 to 80.0, 1.0 to 93.0, 2.5 to 100.0))

    /**
     * Volume of the last 7 days divided by the weekly average of the last 28. Around 1 is
     * sustainable; a sharp spike is the classic overreaching sign.
     */
    fun trainingLoad(ratio: Double): Double =
        interpolate(ratio, listOf(0.0 to 70.0, 0.5 to 82.0, 0.8 to 92.0, 1.0 to 88.0, 1.3 to 75.0, 1.6 to 45.0, 2.5 to 20.0))

    /** Linear interpolation between the given points, flat outside their range. */
    fun interpolate(x: Double, points: List<Pair<Double, Double>>): Double {
        val first = points.firstOrNull() ?: return NEUTRAL
        val last = points.last()
        if (x <= first.first) return first.second
        if (x >= last.first) return last.second
        for ((lower, upper) in points.zipWithNext()) {
            if (!(x <= upper.first)) continue
            val span = upper.first - lower.first
            if (!(span > 0)) return upper.second
            return lower.second + (upper.second - lower.second) * (x - lower.first) / span
        }
        return last.second
    }
}

/**
 * A measurement together with the personal baseline it is judged against. Unit-agnostic: the
 * z-score compares like with like, so any HRV flavour (SDNN, RMSSD) works if value and history
 * share it.
 */
data class MetricBaseline(
    val value: Double,
    val mean: Double,
    val standardDeviation: Double,
) {
    /**
     * Deviation in standard deviations, clamped to ±3. `null` when the baseline barely varies,
     * where a z-score would explode.
     */
    val zScore: Double?
        get() {
            if (!(standardDeviation > mean * 0.01 && standardDeviation > 0)) return null
            return minOf(maxOf((value - mean) / standardDeviation, -3.0), 3.0)
        }

    /** Deviation from the baseline as a fraction, for the human-readable line. */
    val relativeDeviation: Double
        get() {
            if (mean == 0.0) return 0.0
            return (value - mean) / mean
        }

    companion object {
        /**
         * Baseline over earlier daily values, judged against the most recent one. `null` below
         * `minimumDays` of history, where "your normal" is guesswork. (Swift failable `init`.)
         */
        fun fromHistory(current: Double, earlierDailyValues: List<Double>, minimumDays: Int = 10): MetricBaseline? {
            if (earlierDailyValues.size < minimumDays) return null
            val count = earlierDailyValues.size.toDouble()
            val mean = earlierDailyValues.fold(0.0) { sum, v -> sum + v } / count
            val variance = earlierDailyValues.fold(0.0) { sum, v -> sum + (v - mean) * (v - mean) } / count
            return MetricBaseline(value = current, mean = mean, standardDeviation = sqrt(variance))
        }
    }
}
